llpc/tool/llpcPipelineBuilder.cpp

```diff
@@ -143,5 +144,31 @@ void PipelineBuilder::printPipelineInfo(PipelineBuildInfo buildInfo) {
                                   << "\n");
 }
 
+// =====================================================================================================================
+// Output LLPC resulting binary (ELF binary, ISA assembly text, or LLVM bitcode) to the specified target file.
+//
+// @param pipelineBin : Output elf pipeline binary
+// @param suppliedOutFile : Name of the file to output ELF binary (specify "" to use base name of first input file with
+// appropriate extension; specify "-" to use stdout)
+// @param firstInFile : Name of first input file
+// @param index : index of elf binaries
```

Comment:
I don't understand this comment. This function outputs a single elf, so what is an index of (multiple) binaries? I guess this may be the index of the compiled pipeline or the output pipeline? IE for existing pipeline types this will be 0, and for new pipeline types that produce multiple Elfs per pipeline this can be more than 0?

Could you update the description to make this clear?

Reply:
Or actually, maybe it'd be better to require callers to make sure the name is never empty? Then we can remove the index altogether and don't need to add unit tests in this PR.

Reply:
if (outFileName != "-" && index > 0) {
    outFileName.append(".");
    outFileName.append(utostr(index));
  }

Reply:
this is index and binaries are mostly for internal project to reduce merge conflict. it can not be arbitrary BinaryData or index, has to be one of pipeline binaries.

Reply:
OK, then I think it's better to move this entirely to each derived class and remove the index parameter from the helper. We can achieve the same thing by calculating the final filename in the caller.

Reply:
If we remove the index , how to reuse the help function "OutputElf"?

For multiple pipeline binaries
for (unsigned i = 0; i < pipelineBinCount; ++i) {
      const BinaryData &pipelineBin = compileInfo->xxxPipelineOut.pipelineBins[i];
      outputElf(pipelineBin, i);
}

for simple pipeline binary
outputElf(compileInfo->gfxPipelineOut.pipelineBins[i], 0);

Reply:
okay, i got your ideas. i will think about it

Reply:
Ah, I think that at this point we may just use `writeFile` directly instead of `outputElf` in `outputElfs`. I think that's reasonable. There is not that much logic to reuse anyway. If extension replacement is useful, it can become its own helper function, or we can just keep it inline.